cmd/access/cluster.go

```diff
@@ -0,0 +1,446 @@
+			cmdutil.CheckErr(verifyPermissions(streams, flags))
+			client := k8s.NewClient(flags)
+			clusterAccess := newClusterAccessOptions(client, streams, flags)
+			cmdutil.CheckErr(clusterAccess.Run(cmd, args))
+		},
+	}
+	return clusterCmd
+}
+
+// clusterCmdComplete verifies the command's invocation, returning an error if the usage is invalid
+func clusterCmdComplete(cmd *cobra.Command, args []string) error {
+	if len(args) != 1 {
+		return cmdutil.UsageErrorf(cmd, "Exactly one cluster identifier was expected")
+	}
+	return osdctlutil.IsValidClusterKey(args[0])
+}
+
+// verifyPermissions determines if the user has supplied the correct permissions in order to retrieve a KubeConfig secret from hive.
+// If the user attempts to impersonate an unexpected account, an error is returned.
+// If the user hasn't attempted to impersonate anyone, it prompts whether they would like to do so automatically.
+func verifyPermissions(streams genericclioptions.IOStreams, flags *genericclioptions.ConfigFlags) error {
+	if flags.Impersonate != nil && *flags.Impersonate != "" {
+		if *flags.Impersonate != impersonateUser {
+			osdctlutil.StreamPrintln(streams, "")
+			return fmt.Errorf("Unauthorized impersonation as user '%s'. Only requests to impersonate '%s' are allowed.", *flags.Impersonate, impersonateUser)
+		}
+	} else {
+		osdctlutil.StreamPrintln(streams, "")
+		osdctlutil.StreamPrintln(streams, fmt.Sprintf("No impersonation request detected. By design, SREs do not have sufficient permission to retrieve a cluster Kubeconfig from hive, and should impersonate '%s' to do so.", impersonateUser))
+		osdctlutil.StreamPrint(streams, fmt.Sprintf("Would you like to continue as '%s'? (You can disable this prompt in the future by rerunning this command with '--as %s') [y/N] ", impersonateUser, impersonateUser))
+
+		input, err := osdctlutil.StreamRead(streams, '\n')
+		if err != nil {
+			return err
+		}
+		if !isAffirmative(strings.TrimSpace(input)) {
+			return fmt.Errorf("Did not impersonate '%s'", impersonateUser)
+		}
+		*flags.Impersonate = impersonateUser
+		osdctlutil.StreamPrintln(streams, fmt.Sprintf("Continuing as '%s'", impersonateUser))
+		osdctlutil.StreamPrintln(streams, "")
+	}
+	return nil
+}
+
+// clusterAccessOptions contains the objects and information required to access a cluster
+type clusterAccessOptions struct {
+	*genericclioptions.ConfigFlags
+	genericclioptions.IOStreams
+	kclient.Client
+}
+
+// newAccessOptions creates a clusterAccessOptions object
+func newClusterAccessOptions(client kclient.Client, streams genericclioptions.IOStreams, flags *genericclioptions.ConfigFlags) clusterAccessOptions {
+	a := clusterAccessOptions{
+		IOStreams:   streams,
+		ConfigFlags: flags,
+		Client:      client,
+	}
+	return a
+}
+
+// Println appends a newline then prints the given msg using the clusterAccessOptions' IOStreams
+func (c *clusterAccessOptions) Println(msg string) {
```

Comment:
These functions looks like they should be in utils, but I'm not sure I understand why this is being used, could you explain?

Reply:
They're just wrappers for the `Stream*` functions that utilize the `clusterAccessOptions`'s `IOStreams`.

Using `c.Println(msg)` or `c.Print(msg)` is more idiomatic than `osdctlutils.StreamPrintln(c.IOStreams, msg)` or `c.IOStreams.Out.Write([]byte{msg})` for every I/O operation